src/Domain/ZCurve.cpp

```diff
@@ -80,11 +83,176 @@ size_t z_curve_index_from_element_id(const ElementId<Dim>& element_id) {
+  // refinement (y) is 3. i.e. extract only the z2 y2 z1 y1 x1 z0 y0 x0 bits
+  // of the whole Z-curve index, z4 z3 z2 y2 z1 y1 x1 z0 y0 x0.
+  while (starting_dim_index < Dim - 1) {
+    // refinement level of lowest dim we're still extracting bits for
+    const size_t refinement_level =
+        gsl::at(dimension_by_highest_refinement_level, starting_dim_index)
+            .first;
+    // number of bits to extract for each dimension we're still extracting bits
+    // for
+    const size_t num_bits_to_extract = refinement_level - bit_index;
+    for (size_t i = 0; i < num_bits_to_extract; i++) {
+      for (size_t dim_index = starting_dim_index; dim_index < Dim;
+           dim_index++) {
+        // extract lowest bit of what remains of the Z-curve index value, shift
+        // the bit up to the bit position we're on, and add it to the result
+        // segment index
+        segment_indices_by_highest_refinement_level.at(dim_index).first |=
+            ((element_order_index & 1) << bit_index);
+        // clear the lowest bit that we just completed extracting
+        element_order_index >>= 1;
+      }
+      bit_index++;
+    }
+
+    starting_dim_index++;
+    // move the starting dim index to the dim with the next-highest refinement
+    // since we've extracted all bits up to this refinement level for all dims
+    while (starting_dim_index < Dim and
+           refinement_level == gsl::at(dimension_by_highest_refinement_level,
+                                       starting_dim_index)
+                                   .first) {
+      starting_dim_index++;
+    }
+  }
+  // Extract the remaining highest bits of the highest refined dimension that
+  // are not shared by the dimension with the next-highest refinement. For
+  // example, if we have refinement 2 in x, 3 in y, and 5 in z, then all bits of
+  // x and y and only the first 3 bits of z have already been extracted, so this
+  // will just extract the remaining 2 bits for z. i.e. extract only the
+  // remaining z4 z3 bits of the whole Z-curve index,
+  // z4 z3 z2 y2 z1 y1 x1 z0 y0 x0.
+  //
+  // If the highest refinement is shared by more than one dimension, this will
+  // have no effect since all bits have already been extracted in the loop
+  // above.
+  segment_indices_by_highest_refinement_level.at(Dim - 1).first |=
+      (element_order_index << bit_index);
+
+  // result segment indices reordered with the dimension ordering of the input,
+  // i.e. undo the initial sorting of dimension by highest refinement
+  std::array<size_t, Dim> result_segment_indices{};
+  for (size_t i = 0; i < Dim; i++) {
+    gsl::at(result_segment_indices,
+            gsl::at(segment_indices_by_highest_refinement_level, i).second) =
+        gsl::at(segment_indices_by_highest_refinement_level, i).first;
+  }
+
+  return result_segment_indices;
+}
+
+template <>
+std::vector<ElementId<1>> initial_element_ids_in_z_curve_order<1>(
```

Comment:
I almost did this instead but then didn't just because I thought the combined one looked ugly with `if constexpr` inside the for loop 😂 but I don't have a strong opinion either way and will update it to the way you're describing

Reply:
"Always" prefer `if constexpr`. It's easier to read and reason about than specializations. The quotes is if the entire function body is different then specializations are better